File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;
import java.io.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        
        int onBridgeW = 0;
        Queue<Integer> q = new ArrayDeque<>();
        for(int i=0 ; i<bridge_length; i++){
            q.offer(0);
        }
        
        int idx = 0;
        for(int t=1; ; t++){
            // 하차
            if(!q.isEmpty()){
                int w = q.poll();
                onBridgeW -= w;
            }
            
            // 종료
            if(idx == truck_weights.length && onBridgeW == 0){
                answer = t;
                break;
            }
            
            
            // 상차
            if(idx < truck_weights.length && 
               onBridgeW + truck_weights[idx] <= weight){
                onBridgeW += truck_weights[idx];
                q.offer(truck_weights[idx]);
                idx++;
            }else{
                q.offer(0);
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    
    int onBridgeW = 0;
    queue<int> q; // 다리 위 트럭들 관리, 다리 길이만큼 큐 길이 유지
    for(int i=0 ; i<bridge_length; i++){
        q.push(0);
    }
    
    for(int t=1; ; t++){
        
        // 하차
        if(!q.empty()){
            int topW = q.front();
            q.pop();
            onBridgeW -= topW;
        }
        
        // 종료
        if(onBridgeW == 0 && truck_weights.size() == 0){
            answer = t;
            break;
        }
        
        // 상차
        if(truck_weights.size() > 0 && (onBridgeW + truck_weights[0] <= weight)){
            q.push(truck_weights[0]);
            onBridgeW += truck_weights[0];
            truck_weights.erase(truck_weights.begin());
        }else{
            q.push(0);
        }
    }
    
    
    return answer;
}